keps/sig-storage/0035-20190122-inline-csi-volumes-support.md

```diff
@@ -0,0 +1,350 @@
+---
+kep-number: 35
+title: Inline CSI Volumes Support
+authors:
+  - "@vladimirvivien"
+owning-sig: sig-storage
+participating-sigs:
+  - sig-aaa
+  - sig-bbb
+reviewers:
+  - "@thockin"
+  - "@saad-ali"
+  - "@msau42"
+  - "@jsafrane"
+approvers:
+  - TBD
+creation-date: 2019-01-22
+status: implementable
+---
+
+# Inline CSI volumes support
+
+## Table of Contents
+
+* [Table of Contents](#table-of-contents)
+* [Summary](#summary)
+* [Motivation](#motivation)
+    * [Goals](#goals)
+    * [Non-Goals](#non-goals)
+* [User Stories](#user-stories)
+    * [Examples](#examples)
+* [Proposal](#proposal)
+    * [VolumeSource API](#volumesource-api)
+    * [Secret references](#secret-references)
+    * [Specifying allowed inline drivers with PodSecurityPolicy](#specifying-allowed-inline-drivers-with-podSecuritypolicy)
+    * [Persistent vs ephemeral lifecycle setting](#persistent-vs-ephemeral-lifecycle-setting)
+    * [VolumeHandle generation](#volumehandle-generation)
+    * [Inline CSI volume operation stages](#inline-csi-volume-operation-stages)
+    * [Security considerations](#security-considerations)
+
+## Summary
+Currently, volumes that are backed by CSI drivers can only be used with the `PersistentVolume` and `PersistentVolumeClaim` objects. This proposal is to implement support for the ability to nest CSI volume declarations within pod specs, as is supported already by existing in-tree storage providers.
+
+This KEP started life as [feature #2273](https://github.com/kubernetes/community/pull/2273).  Please follow that link for historical context.
+
+
+## Motivation
+Implementing support for embedding volumes directly in pod specs would allow external CSI drivers to have feature parity with their in-tree counter parts.  There are several reasons why this is desirable:
+* It provides the community a path to move away from in-tree volume plugins to CSI, as designed in a separate proposal https://github.com/kubernetes/community/pull/2199/. 
+* This feature would make it possible to create new types of CSI drivers such as ephemeral volume drivers.  They can be used to inject arbitrary states such as configuration, secrets, identity, variables or similar information inside pods. 
+* This can also provide a migration path for older Flex-style drivers (allowing the deprecation of Flex.)
+
+### Goals 
+* Provide a high level design for inline CSI volumes support
+* Define API changes needed to support this feature
+* Inline CSI volumes should work with existing CSI drivers
+* Design for ephemeral inline CSI volumes support
+* Ensure that inline CSI volumes usage is secure
+
+### Non-goals
+The followings will not be addressed by this KEP:
+* Introduce new CSI spec changes to support this feature
+* Introduce required changes to existing CSI drivers for this feature
+* Implementation of ephemeral inline support
+* Support for topology or pod placement scheme for ephemeral inline volumes
```

Comment:
what about other features that only work with pvs, like raw block, mount options, resizing, etc

Reply:
@msau42 this KEP focuses on the minimum features to provide a path for feature parity with in-tree drivers. Additional features will be better served in future KEP.

Reply:
Yup, so we should call it out as a non-goal to support all these other features